backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+	isCreated := false
+	isCompleted := false
+	argIndex := 1
+
+	for _, e := range wi.State.NewEvents() {
+		if es := e.GetExecutionStarted(); es != nil {
+			if isCreated {
+				// TODO: Log warning about duplicate start event
+				continue
+			}
+			isCreated = true
+			sqlSB.WriteString(fmt.Sprintf("CreatedTime = $%d, Input = $%d, ", argIndex, argIndex+1))
+			sqlUpdateArgs = append(sqlUpdateArgs, e.Timestamp.AsTime())
+			sqlUpdateArgs = append(sqlUpdateArgs, es.Input.GetValue())
+			argIndex += 2
+		} else if ec := e.GetExecutionCompleted(); ec != nil {
+			if isCompleted {
+				// TODO: Log warning about duplicate completion event
+				continue
+			}
+			isCompleted = true
+			sqlSB.WriteString(fmt.Sprintf("CompletedTime = $%d, Output = $%d, FailureDetails = $%d, ", argIndex, argIndex+1, argIndex+2))
+			sqlUpdateArgs = append(sqlUpdateArgs, now)
+			sqlUpdateArgs = append(sqlUpdateArgs, ec.Result.GetValue())
+			if ec.FailureDetails != nil {
+				bytes, err := protojson.Marshal(ec.FailureDetails)
+				if err != nil {
+					return fmt.Errorf("failed to marshal FailureDetails: %w", err)
+				}
+				sqlUpdateArgs = append(sqlUpdateArgs, &bytes)
+			} else {
+				sqlUpdateArgs = append(sqlUpdateArgs, nil)
+			}
+			argIndex += 3
+		}
+		// TODO: Execution suspended & resumed
+	}
+
+	if wi.State.CustomStatus != nil {
+		sqlSB.WriteString(fmt.Sprintf("CustomStatus = $%d, ", argIndex))
+		sqlUpdateArgs = append(sqlUpdateArgs, wi.State.CustomStatus.Value)
+		argIndex++
+	}
+
+	// TODO: Support for stickiness, which would extend the LockExpiration
+	sqlSB.WriteString(fmt.Sprintf(
+		"RuntimeStatus = $%d, LastUpdatedTime = $%d, LockExpiration = NULL WHERE InstanceID = $%d AND LockedBy = $%d",
+		argIndex, argIndex+1, argIndex+2, argIndex+3))
+	sqlUpdateArgs = append(sqlUpdateArgs, helpers.ToRuntimeStatusString(wi.State.RuntimeStatus()), now, string(wi.InstanceID), wi.LockedBy)
+	argIndex += 4
+
+	command := sqlSB.String()
+	result, err := tx.ExecContext(ctx, command, sqlUpdateArgs...)
+	if err != nil {
+		return fmt.Errorf("failed to update Instances table: %w", err)
+	}
+
+	count, err := result.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("failed to get the number of rows affected by the Instance table update: %w", err)
+	} else if count == 0 {
+		return fmt.Errorf("instance '%s' no longer exists or was locked by a different worker", string(wi.InstanceID))
+	}
+
+	// If continue-as-new, delete all existing history
+	if wi.State.ContinuedAsNew() {
+		if _, err := tx.ExecContext(ctx, "DELETE FROM durabletask.History WHERE InstanceID = $1", string(wi.InstanceID)); err != nil {
+			return fmt.Errorf("failed to delete from History table: %w", err)
+		}
+	}
+
+	// Save new history events
+	newHistoryCount := len(wi.State.NewEvents())
+	if newHistoryCount > 0 {
+		var sqlSB strings.Builder
+		sqlSB.WriteString("INSERT INTO durabletask.History (InstanceID, SequenceNumber, EventPayload) VALUES ($1, $2, $3)")
```

Comment:
One of the best things about pgx is the CopyFrom method. Allows inserting multiple rows at once much more efficiently than having to build a query like this: https://pkg.go.dev/github.com/jackc/pgx/v5#hdr-Copy_Protocol